src/object/cli_obj.c

```diff
@@ -447,23 +450,35 @@ obj_grp_valid_shard_get(struct dc_object *obj, int idx,
 		return -DER_STALE;
 	}
 
-	idx_first = (idx / grp_size) * grp_size;
-	idx = idx_first + random() % grp_size;
+	D_DEBUG(DB_IO, "grp size %d replicas %d\n", grp_size,
+		obj_get_replicas(obj));
+	/* Start from an random offset within this group, NB: we should
+	 * use replica number directly, instead of group size, which might
+	 * included extended shard, see pl_map_extend().
+	 */
+	D_ASSERT(grp_size >= obj_get_replicas(obj));
+	grp_start = grp_idx * grp_size;
+	idx = grp_start + random() % obj_get_replicas(obj);
```

Comment:
yes, obj_get_replicas will always return the object class replica here.

Reply:
So without the extended shard, right?

Reply:
Is it possible that some object may have multiple redundancy groups, but different redundancy groups have different real shards count because of OSA? If that happened, whether it will affect the shard/leader election?

Reply:
oh, all group will have the same shard cnt, even it is not needed to be added.

Reply:
Assume that the object has two redundancy groups with three replicas. Originally, without OSA case, its total shards are: 3 * 2 = 6.

G_0: shard_0, shard_1, shard_2.
G_1: shard_3, shard_4, shard_5.

But at some time point, OSA happen, the G_0 is affected, but G_1 is NOT affected, then the real shards may become:

G_0: shard_0 (old), shard_1 (old), shard_2 (old), shard_3 (extended), shard_4 (extended).
G_1: shard_5 (old), shard_6 (old), shard_7 (old).

Is it a reasonable layout for the object with 3 + 2 + 3 = 8 shards under OSA case? Or the G_1 will also has to become "3 old shards + 2 extended shards"?

Under the OSA case, obj_get_replicas() will return 3 or 5?

According to current obj_get_replicas() implementation, it may return "return obj->cob_grp_size".

Reply:
oh, the shards on each group will always be same. so in your case, it will be
G_0 : shard_0, shard_1, shard_2, new_shard
G_1 : shard_3, shard_4, shard_5, -1